README.md
# koop-provider-elastic-sql
Provider to fetch data in  elastic/opensearch datastores through the SQL API.  This provider leverages the SQL API exclusively, and as such, it is limited by the capabilities of that API.

## Data considerations
In order to gain the greatest efficiency from this provider, each document/record in _your dataset should have a unique-indentifier that is a property of the document/record_. This is important, because even though Elastic will create its own `_id` field as part of a document's metadata, _metadata fields are not accessible via the SQL API_.

If your data has a geometry and you want leverage Elastic's geo-filter capability (and you should), the geometry field needs to be mapped in your Elasticsearch Index as a [geopoint](https://www.elastic.co/guide/en/elasticsearch/reference/current/geo-point.html) or [geoshape](https://www.elastic.co/guide/en/elasticsearch/reference/current/geo-shape.html).

A key aspect of working with Elastic is that its API will only return a maximum of 10000 records in any request. If your dataset has more than 10000 records and you request doesn't result in a set that is less than 10000 records, the results will be truncated. This limits the ability to paginate.  To reduce the risk of truncation you should always use a unique-identifier property, a mapped geometry field, and registered the provider with the index mappings for these fields (more on this below).

## Usage

Register the provider with Koop:

```js
const Koop = require('@koopjs/koop-core');
const koop = new Koop({ logLevel: 'info'});
const elasticSqlProvider = require('@koopjs/provider-elastic-sql');

koop.register(elasticSqlProvider, { conn, idFieldMap, geometryFieldMap });
```

### Registration parameters

#### `conn` (required)
An object that contains connection details for the target Elastic cluster.  See the [Elastic documentation](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/client-connecting.html) for details. 

#### `idFieldMap` (optional)
An key/value populated object that serves as a lookup/dictionary for the unique-identifier field of a given Elastic index.  For example, if your index is named "my-index", and the documents in the that index have a field called "some-id-prop" that can be used as a unique-identifier, then the `idFieldMap` would look like:

```js
const idFieldMap = {
  'my-index': 'some-id-prop'
};
```

The `idFieldMap` can have multiple key/value pairs if your Elastic cluster has multiple indicies.

If you omit the `idFieldMap`, filtering requested with the GeoService `objectsIds` parameter will not be executed by the Elastic SQL API.

#### `geometryFieldMap` (optional)
An key/value populated object that serves as a lookup/dictionary for the geopoint/geoshape field of a given Elastic index.  For example, if your index is named "my-index", and the geometry of a document is stored in a geoshape field called "the-geo-prop", then the `geometryFieldMap` would look like:

```js
const idFieldMap = {
  'my-index': 'the-geo-prop'
};
```

While elastic has no limit on the number of geo-fields per document, conversion of a document to GeoJSON requires that we identify a single field as the definitive geometry. The `geometryFieldMap` can have multiple key/value pairs if your Elastic cluster has multiple indicies.

If you omit the `geometryFieldMap`, the GeoJSON produced by the provider will not include a geometry.  In addition, geometry filtering will not be executed by the Elastic instance.

### Route parameters

#### `id`
Once registered with Koop, the provide will expose routes with an `id` parameter. For example:

```sh
/elastic-sql/rest/services/:id/FeatureServer
```

The `id` parameter should be filled with the name of the Elastic index you are targeting.  So if you wished to query the `fires` document-index on you Elastic instance, you would make a request to:

```sh
/elastic-sql/rest/services/fires/FeatureServer/0/query
```


## Demo

The repository includes a demonstration project.  To run the demo you will need Docker installed on your computer. Once installed you can following the steps below to create a local Elastic instance and load it with sample data:

```sh
> npm install

> cd demo

# use Docker to run Elastic/Kibana
> docker-compose up -d

# load sample data; this will create an Elastc index named "fires"
> node loader.js 

# start the Koop application
> node demo.js
```

CHANGELOG.md
# @koopjs/provider-elastic-sql

## 1.0.1
### Major Changes
- initial release

## 1.0.0
### Major Changes
- initial release

src/index.spec.js
const { name, type, Model, version } = require('./index');

describe('registration object', () => {
  test('has expected properties', () => {
    expect(name).toBe('elastic-sql');
    expect(type).toBe('provider');
    expect(version).toBeDefined();
    expect(Model).toBeInstanceOf(Function);
  });
});

src/model.spec.js
const { promisify } = require('util');
const Provider = require('./model');
const elastic = require('@elastic/elasticsearch');

jest.mock('@elastic/elasticsearch');
const elasticQueryMock = jest.fn(() => {
  return {
    rows: [
      ['AK64', 'POINT (-144.67 64.48)', 'GLACIER CREEK'],
      ['SD43', 'POINT (-102.84 43.08)', 'SANDOZ'],
    ],
    columns: [{ name: 'id' }, { name: 'location' }, { name: 'name' }],
  };
});

elastic.Client.mockImplementation(function () {
  return {
    sql: {
      query: elasticQueryMock,
    },
  };
});

const logger = {
  error: () => {},
};

describe('Model', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('constructor', () => {
    test('instantiate without options', () => {
      const provider = new Provider(
        { logger },
        { conn: { node: 'http://localhost' } },
      );
      expect(provider.getData).toBeInstanceOf(Function);
    });

    test('instantiate with options', () => {
      const provider = new Provider(
        { logger },
        {
          conn: { node: 'http://localhost' },
          idFieldMap: { 'my-index': 'id' },
          geometryFieldMap: { 'my-index': 'location' },
        },
      );
      expect(provider.getData).toBeInstanceOf(Function);
    });

    test('instantiate with invalid options', () => {
      try {
        new Provider(
          { logger },
          {
            conn: 'test',
          },
        );
        throw new Error('should have thrown');
      } catch (error) {
        expect(error.message).toBe(
          'invalid "conn", must be of type object',
        );
      }
    });
  });

  describe('getData', () => {
    test('with field maps, no query options', async () => {
      const provider = new Provider(
        { logger },
        {
          conn: { node: 'http://localhost' },
          idFieldMap: { 'my-index': 'id' },
          geometryFieldMap: { 'my-index': 'location' },
        },
      );
      const getData = promisify(provider.getData).bind(provider);

      const result = await getData({ query: {}, params: { id: 'my-index' } });
      expect(result).toEqual({
        type: 'FeatureCollection',
        features: [
          {
            geometry: { type: 'Point', coordinates: [-144.67, 64.48] },
            properties: { id: 'AK64', name: 'GLACIER CREEK' },
          },
          {
            geometry: { type: 'Point', coordinates: [-102.84, 43.08] },
            properties: { id: 'SD43', name: 'SANDOZ' },
          },
        ],
        metadata: { idField: 'id' },
        filtersApplied: {},
      });
      expect(elasticQueryMock.mock.calls[0]).toEqual([
        {
          fetch_size: 1000,
          query: 'SELECT *, location FROM my-index',
        },
      ]);
    });

    test('with field maps, with geometry filter', async () => {
      const provider = new Provider(
        { logger },
        {
          conn: { node: 'http://localhost' },
          idFieldMap: { 'my-index': 'id' },
          geometryFieldMap: { 'my-index': 'location' },
        },
      );
      const getData = promisify(provider.getData).bind(provider);

      const result = await getData({
        query: {
          geometry: '-123,45,-120,49',
        },
        params: { id: 'my-index' },
      });

      expect(result).toEqual({
        type: 'FeatureCollection',
        features: [
          {
            geometry: { type: 'Point', coordinates: [-144.67, 64.48] },
            properties: { id: 'AK64', name: 'GLACIER CREEK' },
          },
          {
            geometry: { type: 'Point', coordinates: [-102.84, 43.08] },
            properties: { id: 'SD43', name: 'SANDOZ' },
          },
        ],
        metadata: { idField: 'id' },
        filtersApplied: {
          geometry: true,
        },
      });
      expect(elasticQueryMock.mock.calls[0]).toEqual([
        {
          fetch_size: 1000,
          query: 'SELECT *, location FROM my-index',
          filter: {
            geo_shape: {
              location: {
                relation: 'intersects',
                shape: {
                  coordinates: [
                    [
                      [-123, 45],
                      [-120, 45],
                      [-120, 49],
                      [-123, 49],
                      [-123, 45],
                    ],
                  ],
                  type: 'polygon',
                },
              },
            },
          },
        },
      ]);
    });

    test('with field maps, with geometry filter w/ relation', async () => {
      const provider = new Provider(
        { logger },
        {
          conn: { node: 'http://localhost' },
          idFieldMap: { 'my-index': 'id' },
          geometryFieldMap: { 'my-index': 'location' },
        },
      );
      const getData = promisify(provider.getData).bind(provider);

      const result = await getData({
        query: {
          geometry: '-123,45,-120,49',
          spatialRel: 'esriSpatialRelUnknown'
        },
        params: { id: 'my-index' },
      });

      expect(result).toEqual({
        type: 'FeatureCollection',
        features: [
          {
            geometry: { type: 'Point', coordinates: [-144.67, 64.48] },
            properties: { id: 'AK64', name: 'GLACIER CREEK' },
          },
          {
            geometry: { type: 'Point', coordinates: [-102.84, 43.08] },
            properties: { id: 'SD43', name: 'SANDOZ' },
          },
        ],
        metadata: { idField: 'id' },
        filtersApplied: {
          geometry: true,
        },
      });

      expect(elasticQueryMock.mock.calls[0]).toEqual([
        {
          fetch_size: 1000,
          query: 'SELECT *, location FROM my-index',
          filter: {
            geo_shape: {
              location: {
                relation: 'intersects',
                shape: {
                  coordinates: [
                    [
                      [-123, 45],
                      [-120, 45],
                      [-120, 49],
                      [-123, 49],
                      [-123, 45],
                    ],
                  ],
                  type: 'polygon',
                },
              },
            },
          },
        },
      ]);
    });

    test('with field maps, with where filter', async () => {
      const provider = new Provider(
        { logger },
        {
          conn: { node: 'http://localhost' },
          idFieldMap: { 'my-index': 'id' },
          geometryFieldMap: { 'my-index': 'location' },
        },
      );
      const getData = promisify(provider.getData).bind(provider);

      const result = await getData({
        query: {
          where: 'foo=\'bar\'',
        },
        params: { id: 'my-index' },
      });

      expect(result).toEqual({
        type: 'FeatureCollection',
        features: [
          {
            geometry: { type: 'Point', coordinates: [-144.67, 64.48] },
            properties: { id: 'AK64', name: 'GLACIER CREEK' },
          },
          {
            geometry: { type: 'Point', coordinates: [-102.84, 43.08] },
            properties: { id: 'SD43', name: 'SANDOZ' },
          },
        ],
        metadata: { idField: 'id' },
        filtersApplied: {
          where: true,
        },
      });
      expect(elasticQueryMock.mock.calls[0]).toEqual([
        {
          fetch_size: 1000,
          query: "SELECT *, location FROM my-index WHERE foo='bar'",
        },
      ]);
    });

    test('with field maps, with objectIds filter', async () => {
      const provider = new Provider(
        { logger },
        {
          conn: { node: 'http://localhost' },
          idFieldMap: { 'my-index': 'id' },
          geometryFieldMap: { 'my-index': 'location' },
        },
      );
      const getData = promisify(provider.getData).bind(provider);

      const result = await getData({
        query: {
          objectIds: 'a,b,c',
        },
        params: { id: 'my-index' },
      });

      expect(result).toEqual({
        type: 'FeatureCollection',
        features: [
          {
            geometry: { type: 'Point', coordinates: [-144.67, 64.48] },
            properties: { id: 'AK64', name: 'GLACIER CREEK' },
          },
          {
            geometry: { type: 'Point', coordinates: [-102.84, 43.08] },
            properties: { id: 'SD43', name: 'SANDOZ' },
          },
        ],
        metadata: { idField: 'id' },
        filtersApplied: {
          objectIds: true,
        },
      });
      expect(elasticQueryMock.mock.calls[0]).toEqual([
        {
          fetch_size: 1000,
          query: 'SELECT *, location FROM my-index WHERE id IN (\'a\',\'b\',\'c\')',
        },
      ]);
    });

    test('with field maps, with numeric objectIds filter', async () => {
      const provider = new Provider(
        { logger },
        {
          conn: { node: 'http://localhost' },
          idFieldMap: { 'my-index': 'id' },
          geometryFieldMap: { 'my-index': 'location' },
        },
      );
      const getData = promisify(provider.getData).bind(provider);

      const result = await getData({
        query: {
          objectIds: '1,2',
        },
        params: { id: 'my-index' },
      });

      expect(result).toEqual({
        type: 'FeatureCollection',
        features: [
          {
            geometry: { type: 'Point', coordinates: [-144.67, 64.48] },
            properties: { id: 'AK64', name: 'GLACIER CREEK' },
          },
          {
            geometry: { type: 'Point', coordinates: [-102.84, 43.08] },
            properties: { id: 'SD43', name: 'SANDOZ' },
          },
        ],
        metadata: { idField: 'id' },
        filtersApplied: {
          objectIds: true,
        },
      });
      expect(elasticQueryMock.mock.calls[0]).toEqual([
        {
          fetch_size: 1000,
          query: 'SELECT *, location FROM my-index WHERE id IN (1,2)',
        },
      ]);
    });

    test('with field maps, with orderByFields sorting', async () => {
      const provider = new Provider(
        { logger },
        {
          conn: { node: 'http://localhost' },
          idFieldMap: { 'my-index': 'id' },
          geometryFieldMap: { 'my-index': 'location' },
        },
      );
      const getData = promisify(provider.getData).bind(provider);

      const result = await getData({
        query: {
          orderByFields: 'name',
        },
        params: { id: 'my-index' },
      });

      expect(result).toEqual({
        type: 'FeatureCollection',
        features: [
          {
            geometry: { type: 'Point', coordinates: [-144.67, 64.48] },
            properties: { id: 'AK64', name: 'GLACIER CREEK' },
          },
          {
            geometry: { type: 'Point', coordinates: [-102.84, 43.08] },
            properties: { id: 'SD43', name: 'SANDOZ' },
          },
        ],
        metadata: { idField: 'id' },
        filtersApplied: {
          orderByFields: true,
        },
      });
      expect(elasticQueryMock.mock.calls[0]).toEqual([
        {
          fetch_size: 1000,
          query: 'SELECT *, location FROM my-index ORDER BY name',
        },
      ]);
    });

    test('without field maps, no query options', async () => {
      const provider = new Provider(
        { logger },
        {
          conn: { node: 'http://localhost' }
        },
      );
      const getData = promisify(provider.getData).bind(provider);

      const result = await getData({ query: {}, params: { id: 'my-index' } });
      expect(result).toEqual({
        type: 'FeatureCollection',
        features: [
          {
            geometry: undefined,
            properties: { id: 'AK64', location: 'POINT (-144.67 64.48)', name: 'GLACIER CREEK' },
          },
          {
            geometry: undefined,
            properties: { id: 'SD43', location: 'POINT (-102.84 43.08)', name: 'SANDOZ' },
          },
        ],
        metadata: { idField: undefined },
        filtersApplied: {},
      });
      expect(elasticQueryMock.mock.calls[0]).toEqual([
        {
          fetch_size: 1000,
          query: 'SELECT * FROM my-index',
        },
      ]);
    });

    test('handle elastic 400 error', async () => {
      elastic.Client.mockImplementationOnce(function () {
        return {
          sql: {
            query: () => {
              const error = new Error('bad input');
              error.name = 'ResponseError';
              error.body = { status: 400 };
              throw error;
            },
          },
        };
      });
      const provider = new Provider(
        { logger },
        {
          conn: { node: 'http://localhost' },
          idFieldMap: { 'my-index': 'id' },
          geometryFieldMap: { 'my-index': 'location' },
        },
      );
      const getData = promisify(provider.getData).bind(provider);

      try {
        await getData({ query: {}, params: { id: 'my-index' } });
        throw new Error('should have thrown');
      } catch (error) {
        expect(error.message).toBe('invalid input');
        expect(error.code).toBe(400);
      }
    });

    test('handle elastic 500 error', async () => {
      elastic.Client.mockImplementationOnce(function () {
        return {
          sql: {
            query: () => {
              const error = new Error('something went wrong');
              error.name = 'ResponseError';
              throw error;
            },
          },
        };
      });
      const provider = new Provider(
        { logger },
        {
          conn: { node: 'http://localhost' },
          idFieldMap: { 'my-index': 'id' },
          geometryFieldMap: { 'my-index': 'location' },
        },
      );
      const getData = promisify(provider.getData).bind(provider);

      try {
        await getData({ query: {}, params: { id: 'my-index' } });
        throw new Error('should have thrown');
      } catch (error) {
        expect(error.message).toBe('something went wrong');
        expect(error.code).toBe(500);
      }
    });

    test('handle other error', async () => {
      elastic.Client.mockImplementationOnce(function () {
        return {
          sql: {
            query: () => {
              throw new Error('something went wrong');
            },
          },
        };
      });
      const provider = new Provider(
        { logger },
        {
          conn: { node: 'http://localhost' },
          idFieldMap: { 'my-index': 'id' },
          geometryFieldMap: { 'my-index': 'location' },
        },
      );
      const getData = promisify(provider.getData).bind(provider);

      try {
        await getData({ query: {}, params: { id: 'my-index' } });
        throw new Error('should have thrown');
      } catch (error) {
        expect(error.message).toBe('something went wrong');
      }
    });
  });
});


src/validate.spec.js
const { validateConn, validateGeometryFieldMap, validateIdFieldMap } = require('./validate');

describe('validation functions', () => {
  describe('validateConn', () => {
    test('validateConn, success', () => {
      expect(validateConn({})).toBe(undefined);
    });
  
    test('validateConn, failure when missing', () => {
      try {
        validateConn();
        throw new Error('should have thrown');
      } catch (error) {
        expect(error.message).toBe('client connection configuration object is required');
      }
    });
  
    test('validateConn, failure when wrong data type', () => {
      try {
        validateConn('string');
        throw new Error('should have thrown');
      } catch (error) {
        expect(error.message).toBe('invalid "conn", must be of type object');
      }
    });
  });

  describe('validateIdFieldMap', () => {
    test('validateIdFieldMap, success', () => {
      expect(validateIdFieldMap({ index: 'field' })).toBe(undefined);
    });
  
    test('validateIdFieldMap, failure when wrong data type', () => {
      try {
        validateIdFieldMap('string');
        throw new Error('should have thrown');
      } catch (error) {
        expect(error.message).toBe('invalid "idFieldMap", must be of type object');
      }
    });
  });

  describe('validateGeometryFieldMap', () => {
    test('validateGeometryFieldMap, success', () => {
      expect(validateGeometryFieldMap({ index: 'field' })).toBe(undefined);
    });
  
    test('validateGeometryFieldMap, failure when wrong data type', () => {
      try {
        validateGeometryFieldMap('string');
        throw new Error('should have thrown');
      } catch (error) {
        expect(error.message).toBe('invalid "geometryFieldMap", must be of type object');
      }
    });
  });
});

demo/loader.js
const fs = require('fs');
const _ = require('lodash');
const zlib = require('zlib');
const { pipeline } = require('stream/promises');
const { Client } = require('@elastic/elasticsearch');
const client = new Client({
  node: 'http://localhost:9200', // Elasticsearch endpoint,
});

async function execute() {
  try {
    await client.indices.delete({ index: 'fires' });
  } catch (err) {
    console.log('no index to delete');
  }

  await client.index({ index: 'fires', document: {}, refresh: true });
  const result = await client.search({ index: 'fires' });

  await client.delete({ index: 'fires', id: result.hits.hits[0]._id });

  await client.indices.putMapping({
    index: 'fires',
    properties: {
      fireId: {
        type: 'keyword',
      },
      fireName: {
        type: 'text',
        fields: {
          keyword: {
            type: 'keyword',
            ignore_above: 256,
          },
        },
      },
      fireType: {
        type: 'keyword',
      },
      acres: {
        type: 'unsigned_long',
      },
      location: {
        type: 'geo_point',
      },
    },
  });

  await pipeline(
    fs.createReadStream('test-data.json.gz'),
    zlib.Unzip(),
    fs.createWriteStream('wildfires.json'),
  );

  const features = require('./wildfires.json');

  const chunks = _.chain(features)
    .map((feature) => {
      const { properties, geometry } = feature;

      return [
        {
          create: { _index: 'fires' },
        },
        {
          fireId: properties.FIRE_ID,
          fireName: properties.FIRE_NAME,
          fireType: properties.FIRE_TYPE,
          acres: properties.ACRES,
          location: geometry,
        },
      ];
    })
    .flatten()
    .chunk(1000)
    .value();

  for (let i = 0; i < chunks.length; i++) {
    const chunk = chunks[i];
    await client.bulk({
      body: chunk,
    });
  }
  return;
}

execute()
  .then(() => {
    process.exitCode = 0;
  })
  .catch((err) => {
    console.error(err);
    process.exitCode = 1;
  });


jest.config.js
/**
 * For a detailed explanation regarding each configuration property, visit:
 * https://jestjs.io/docs/configuration
 */

/** @type {import('jest').Config} */
const config = {
  // Indicates which provider should be used to instrument code for coverage
  coverageProvider: 'v8',

  // A list of reporter names that Jest uses when writing coverage reports
  // coverageReporters: [
  //   "json",
  //   "text",
  //   "lcov",
  //   "clover"
  // ],

  roots: ['src'],

  // The glob patterns Jest uses to detect test files
  testMatch: ['**/*.spec.js'],

  // An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
  testPathIgnorePatterns: ['/node_modules/'],


  coverageDirectory: './coverage',
  testEnvironment: 'node',
  // The regexp pattern or array of patterns that Jest uses to detect test files
  // testRegex: [],

  // This option allows the use of a custom results processor
  // testResultsProcessor: undefined,
};

module.exports = config;


src/validate.js
const joi = require('joi');

const connSchema = joi.object().required();

const idFieldMapSchema = joi.object().required();

const geometryFieldMapSchema = joi.object().required();

function validateConn(conn) {
  const { error } = connSchema.validate(conn);
  if (error?.message === '"value" is required') {
    throw new Error('client connection configuration object is required');
  }

  if (error) {
    throw new Error(
      `invalid "conn", ${error.details[0].message.replace('"value" ', '')}`,
    );
  }
}

function validateIdFieldMap(idFieldMap) {
  const { error } = idFieldMapSchema.validate(idFieldMap);
  if (error) {
    throw new Error(`invalid "idFieldMap", ${error.details[0].message.replace('"value" ', '')}`);
  }
}

function validateGeometryFieldMap(geometryFieldMap) {
  const { error } = geometryFieldMapSchema.validate(geometryFieldMap);
  if (error) {
    throw new Error(`invalid "geometryFieldMap", ${error.details[0].message.replace('"value" ', '')}`);
  }
}

module.exports = {
  validateConn,
  validateGeometryFieldMap,
  validateIdFieldMap,
};


src/model.js
const { Client } = require('@elastic/elasticsearch');
const _ = require('lodash');
const { wktToGeoJSON } = require('@terraformer/wkt');
const { standardizeGeometryFilter } = require('@koopjs/geoservice-utils');
const {
  validateConn,
  validateIdFieldMap,
  validateGeometryFieldMap,
} = require('./validate');

const relationLookup = {
  esriSpatialRelIntersects: 'intersects',
  esriSpatialRelContains: 'contains',
  esriSpatialRelWithin: 'within',
};

class Model {
  #client;
  #logger;
  #geometryFieldMap;
  #idFieldMap;

  constructor({ logger }, { conn, geometryFieldMap, idFieldMap } = {}) {
    this.#logger = logger;

    // Validate registration options
    try {
      validateConn(conn);

      if (idFieldMap) {
        validateIdFieldMap(idFieldMap);
      }

      if (geometryFieldMap) {
        validateGeometryFieldMap(geometryFieldMap);
      }
    } catch (error) {
      this.#logger.error(`Provider registration error: ${error.message}`);
      throw error;
    }

    // configure the elastic client
    this.#client = new Client(conn);

    // define a lookup to find the geometry field for a given elastic index
    this.#geometryFieldMap = geometryFieldMap || {};

    // define a lookup to find the id-field for a given elastic index
    this.#idFieldMap = idFieldMap || {};
  }

  async getData(req, callback) {
    const {
      query: geoserviceParams,
      params: { id: elasticIndexName },
    } = req;

    const { geometry, resultRecordCount } = geoserviceParams;
    const idField = this.#idFieldMap[elasticIndexName];
    const geometryField = this.#geometryFieldMap[elasticIndexName];

    const dataStoreQueryBody = {
      fetch_size: resultRecordCount || 1000,
    };

    try {
      dataStoreQueryBody.query = buildSqlQuery({
        ...geoserviceParams,
        idField,
        geometryField,
        elasticIndexName,
      });

      if (geometryField && geometry) {
        dataStoreQueryBody.filter = buildGeoshapeFilter({
          ...geoserviceParams,
          geometryField,
        });
      }

      const results = await this.#client.sql.query(dataStoreQueryBody);

      const geojson = convertToGeojson(results, geometryField);

      geojson.metadata = { idField };

      geojson.filtersApplied = generateFiltersApplied({
        ...geoserviceParams,
        idField,
        geometryField,
        geometry,
      });

      callback(null, geojson);
    } catch (err) {
      const error = this.#handleError(err);
      callback(error);
    }
  }

  #handleError(error) {
    const messagePrefix = 'Provider error:';

    if (error.name === 'ResponseError') {
      this.#logger.error(
        `${messagePrefix} data-store query failure, ${error.message}`,
      );
      const message = error?.body?.status === 400 ? 'invalid input' : error.message;
      const err = new Error(message);
      err.code = error?.body?.status || 500;
      return err;
    }

    this.#logger.error(`${messagePrefix}, ${JSON.stringify(error)}`);
    return error;
  }
}

function buildSqlQuery(params) {
  const {
    where,
    outFields = '*',
    orderByFields,
    objectIds,
    geometryField,
    elasticIndexName,
    idField,
  } = params;

  const from = ` FROM ${elasticIndexName}`;

  const selectFields = geometryField
    ? `${outFields}, ${geometryField}`
    : outFields;

  const whereClause = buildSqlWhere({ where, objectIds, idField });

  const orderByClause = orderByFields ? ` ORDER BY ${orderByFields}` : '';

  return `SELECT ${selectFields}${from}${whereClause}${orderByClause}`;
}

function buildSqlWhere({ where, objectIds, idField }) {
  const sqlWhereComponents = [];

  if (!where && objectIds === undefined) {
    return '';
  }

  if (where) {
    sqlWhereComponents.push(where);
  }

  if (idField && objectIds) {
    const objectIdsComponent = objectIds
      .split(',')
      .map((val) => {
        return isNaN(val) ? `'${val}'` : val;
      })
      .join(',')
      .replace(/^/, `${idField} IN (`)
      .replace(/$/, ')');

    sqlWhereComponents.push(objectIdsComponent);
  }

  return ' WHERE ' + sqlWhereComponents.join(' AND ');
}

function buildGeoshapeFilter({ geometryField, geometry, inSR, spatialRel }) {
  const { geometry: geometryFilter, relation } = standardizeGeometryFilter({
    geometry,
    inSR,
    reprojectionSR: 4326,
    spatialRel,
  });

  return {
    geo_shape: {
      [geometryField]: {
        shape: {
          type: geometryFilter.type.toLowerCase(),
          coordinates: geometryFilter.coordinates,
        },
        relation: relationLookup[relation] || 'intersects',
      },
    },
  };
}

function convertToGeojson(results, geometryField) {
  const keys = results.columns.map((col) => col.name);
  const rowToFeature = convertRowToFeature.bind(null, keys, geometryField);
  const features = results.rows.map(rowToFeature);

  return {
    type: 'FeatureCollection',
    features,
  };
}

function convertRowToFeature(featureAttributeKeys, geometryField, row) {
  const properties = row.reduce((feature, val, i) => {
    feature[featureAttributeKeys[i]] = val;
    return feature;
  }, {});

  const geomFieldIndex = featureAttributeKeys.indexOf(geometryField);

  return {
    properties: _.omit(properties, geometryField),
    geometry: row[geomFieldIndex] && wktToGeoJSON(row[geomFieldIndex]),
  };
}

function generateFiltersApplied({ where, objectIds, orderByFields, idField, geometry, geometryField }) {
  const filtersApplied = {};

  if (where) {
    filtersApplied.where = true;
  }

  if (objectIds && idField) {
    filtersApplied.objectIds = true;
  }

  if (orderByFields) {
    filtersApplied.orderByFields = true;
  }

  if (geometry && geometryField) {
    filtersApplied.geometry = true;
  }

  return filtersApplied;
}

module.exports = Model;
